packages/dd-trace/src/appsec/remote_config/manager.js

```diff
@@ -256,16 +250,42 @@ class RemoteConfigManager extends EventEmitter {
 
       // in case the item was already handled by kPreUpdate
       if (item.apply_state === UNACKNOWLEDGED || action === 'unapply') {
-        try {
-          // TODO: do we want to pass old and new config ?
-          const hadListeners = this.emit(item.product, action, item.file, item.id)
-
-          if (hadListeners) {
-            item.apply_state = ACKNOWLEDGED
+        const handler = this._handlers.get(item.product)
+
+        if (handler) {
+          try {
+            if (supportsAckCallback(handler)) {
+              // If the handler accepts an `ack` callback, expect that to be called and set `apply_state` accordinly
+              // TODO: do we want to pass old and new config ?
+              handler(action, item.file, item.id, (err) => {
+                if (err) {
+                  item.apply_state = ERROR
+                  item.apply_error = err.toString()
+                } else if (item.apply_state !== ERROR) {
+                  item.apply_state = ACKNOWLEDGED
+                }
```

Comment:
could do with some code dedup here, we set handler errors for apply states in 3 different places (cb, promise, catch). A function like this:
```javascript
setApplyState (item, err) {
  if (item.apply_state !== UNACKNOWLEDGED) return

  if (err) {
    item.apply_state = ERROR
    item.apply_error = err.toString()
  } else {
    item.apply_state = ACKNOWLEDGED
  }
}
```

Reply:
Good idea. Btw, is there a specific reason you used `err.toString()` instead of `err.message`?

Reply:
No super specific reason, but toString() is supposed to be the standard interface that returns how the object knows it wants to serialize itself, by picking a specific field, you don't allow those errors to decide themselves. Any reason you want to use message instead ?

Reply:
`err.message` is the standard way to get the error message if that's all we care about. If we want to allow the error to give us more than the error message, then we can of course use `toString()`. I just fear that some errors might return too much data with the `toString()` call and all we care about is just the error message itself. But I'm fine either way 🤷

Reply:
Hmm after implementing your suggested change, I'm not sure I like it as much as I thought I would. You can see the diff here:

<img width="877" alt="image" src="https://github.com/user-attachments/assets/c41cfd05-33b5-4d27-8032-d16ea16be0cc">

I think it's harder to read and it's actually one line more than the original implementation. However, I still agree that this big nested if-block isn't nice. So I would rather want to move the entire block somewhere else and just replace it with a function call, eg:

![image](https://github.com/user-attachments/assets/843c4cdc-5760-4777-b4ba-22af0dd72525)